flink-python/pyflink/ml/api/param/base.py

```diff
@@ -0,0 +1,357 @@
+        :return: return this Params.
+        """
+        if other_params is not None:
+            self._param_map.update(other_params._param_map)
+        return self
+
+    def clone(self) -> 'Params':
+        """
+        Creates and returns a deep clone of this Params.
+
+        :return: a clone of this Params.
+        """
+        new_params = Params()
+        new_params._param_map.update(self._param_map)
+        return new_params
+
+
+class ParamInfo(object):
+    """
+    Definition of a parameter, including name, description, type_converter and so on.
+    """
+
+    def __init__(self, name, description, is_optional=True,
+                 has_default_value=False, default_value=None,
+                 type_converter=None):
+        self.name = str(name)
+        self.description = str(description)
+        self.is_optional = is_optional
+        self.has_default_value = has_default_value
+        self.default_value = default_value
+        self.type_converter = TypeConverters.identity if type_converter is None else type_converter
+
+    def __str__(self):
+        return self.name
+
+    def __repr__(self):
+        return "Param(name=%r, description=%r)" % (self.name, self.description)
+
+    def __hash__(self):
+        return hash(str(self.name))
+
+    def __eq__(self, other):
+        if isinstance(other, ParamInfo):
+            return self.name == other.name
+        else:
+            return False
+
+
+class TypeConverters(object):
+    """
+    Factory methods for common type conversion functions for `Param.typeConverter`.
+    The TypeConverter makes PyFlink ML pipeline support more types of parameters. For example,
+    a list could be a list, a range or an array. Validation can also be done in the converters.
+    """
+
+    @staticmethod
+    def _is_numeric(value):
+        vtype = type(value)
+        return vtype in [int, float] or vtype.__name__ == 'long'
+
+    @staticmethod
+    def _is_integer(value):
+        return TypeConverters._is_numeric(value) and float(value).is_integer()
+
+    @staticmethod
+    def _can_convert_to_list(value):
+        vtype = type(value)
+        return vtype in [list, tuple, range, array.array]
+
+    @staticmethod
+    def _can_convert_to_string(value):
+        return isinstance(value, str)
+
+    @staticmethod
+    def identity(value):
+        """
+        Dummy converter that just returns value.
+        """
+        return value
+
+    @staticmethod
+    def to_list(value):
+        """
+        Convert a value to a list, if possible.
+        """
+        if isinstance(value, list):
+            return value
+        elif type(value) in [tuple, range, array.array]:
+            return list(value)
+        else:
+            raise TypeError("Could not convert %s to list" % value)
+
+    @staticmethod
+    def to_list_float(value):
+        """
+        Convert a value to list of floats, if possible.
+        """
+        if TypeConverters._can_convert_to_list(value):
```

Comment:
this seems redundant since `to_list` will also raise `TypeError`?
do we really need the `_can_convert_to_list` internal func?

Reply:
Good idea. I think we don't need the `_can_convert_to_list`.